proto/src/determined/modelmetrics/v1/modelmetrics.proto

```diff
@@ -0,0 +1,85 @@
+syntax = "proto3";
+
+package determined.modelmetrics.v1;
+option go_package = "github.com/determined-ai/determined/proto/pkg/modelmetricsv1";
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "protoc-gen-swagger/options/annotations.proto";
+
+// The current state of the metrics.
+enum State {
+  // The state of the metrics is unknown.
+  STATE_UNSPECIFIED = 0;
+  // The metrics is in an active state.
+  STATE_ACTIVE = 1;
+  // The metrics is persisted to checkpoint storage.
+  STATE_COMPLETED = 2;
+  // The metrics errored.
+  STATE_ERROR = 3;
+  // The metrics has been deleted.
+  STATE_DELETED = 4;
+}
+
+// TrainingMetrics represents model training metrics.
+message TrainingMetrics {
+  option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+    json_schema: {
+      required: [
+        "experiment_id",
+        "trial_id",
+        "step_id",
+        "start_batch",
+        "end_batch",
+        "state"
+      ]
+    }
+  };
+  // The UUID.
+  string uuid = 1;
+  // The ID of the experiment that created this.
+  int32 experiment_id = 2;
+  // The ID of the trial that created this.
+  int32 trial_id = 3;
+  // The ID of the step in the trial.
+  int32 step_id = 4;
+  // The state.
+  State state = 5;
+  // Start batch number of training metrics, equivalent to prior processed
+  // batches for checkpoints and validations.
+  int32 start_batch = 6;
+  // End batch number of training metrics, equivalent to total batches
+  // for checkpoints and validations.
+  int32 end_batch = 7;
+  // Timestamp when this began being saved to persistent storage.
+  google.protobuf.Timestamp start_time = 8;
+  // Timestamp when this completed being saved to persistent storage.
+  google.protobuf.Timestamp end_time = 9;
+  // Metrics calculated on the training set.
+  google.protobuf.Struct metrics = 10;
```

Comment:
Looking at a normal step from `det t describe --metrics --json 7 | jq ".steps[0]"` on my cluster, I see:
```json
{
  "id": 1,
  "state": "COMPLETED",
  "metrics": {
    "num_inputs": 8,
    "avg_metrics": {
      "lr": 9.999999974752427e-07,
      "loss": 2668.724609375,
      "accuracy": 0
    },
    "batch_metrics": [
      {
        "lr": 9.999999974752427e-07,
        "loss": 5476,
        "accuracy": 0
      },
      {
        "lr": 9.999999974752427e-07,
        "loss": 4986.25732421875,
        "accuracy": 0
      },
      {
        "lr": 9.999999974752427e-07,
        "loss": 2601.38232421875,
        "accuracy": 0
      },
      {
        "lr": 9.999999974752427e-07,
        "loss": 2668.724609375,
        "accuracy": 0
      }
    ]
  },
  "end_time": "2021-02-20T00:37:20.441492+00:00",
  "trial_id": 7,
  "checkpoint": null,
  "start_time": "2021-02-20T00:37:14.777461+00:00",
  "validation": null,
  "num_batches": 4,
  "total_batches": 4,
  "prior_batches_processed": 0
}
```

In particular, the `metrics` object at the step level consists of the subfields:
* `num_inputs`
* `avg_metrics`
* `batch_metrics`

I don't think that e.g. our webui or the hpviz project would behave very nicely if the metrics were not packed that way.

Does it make sense to split these things out so we can ensure that we pack them that way in the database?  For example, the rest API could take `num_inputs`, `avg_metrics`, and `batch_metrics` separately, with e.g. `batch_metrics` being optional (I don't think we use those anywhere in the webui or backend, but @mackrorysd or @katport could probably correct me if I'm wrong on that).

Reply:
batch_metrics being the individual batch datapoints from which avg_metrics is aggregated for each step, right? If so, no we're not using those for HP Viz. We use avg_metrics from training and validation metrics as-they-are.

Reply:
I think `num_inputs` is totally unnecessary since the data itself contains the starting batch number and end batch number. And `avg_metrics` should be renamed to `reduced_metrics`. 

> Does it make sense to split these things out so we can ensure that we pack them that way in the database?

In other parts of the system, we put these three things in one object. It's better to be consistent. Also, I think the current schema is more flexible.

Reply:
I think `num_inputs` is important because it's really about records, not batches, and you can't count that from the master side accurately.

Reply:
@rb-determined-ai  I still don't understand why `num_inputs` is needed on the master. Can you explain that?

Reply:
@rb-determined-ai I keep what the current state as the master is. Maybe it makes more sense if we expose the APIs to the users in the future.

Reply:
It's just useful information for visualizations and such.  It's not needed for the experiment/trial/searcher actors or anything.

Reply:
I agree with ryan, without `num_inputs` we will try to infer it from batches and we will be wrong.

Reply:
@rb-determined-ai @stoksc 

`num_inputs` is actually inferred across the codebase based on a wrong assumption that all batches have the same size.

To me, I don't think having `num_inputs` is useful or used at all. Users might want to pack any arbitrary fields as they want if we support initializing metrics in the harness. The packing schema that doesn't make the visualizations work well sound like bugs on the WebUI and CLI side.

This PR doesn't aim at providing any user-facing APIs, fixing any wrong assumptions, or adding any new constraints to packing metrics. Before we have a concrete consensus on what should be done, I won't change any behavior in this PR.

Reply:
> `num_inputs` is actually inferred across the codebase based on a wrong assumption that all batches have the same size.

Yeah, it is master side but it isn't harness side where this comes from right?